wisp-sampling/src/test/kotlin/wisp/sampling/RateLimiterTest.kt

```diff
@@ -0,0 +1,193 @@
+        assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isTrue()
+        assertThat(ticker.nowMs).isEqualTo(0L)
+
+        // Exhausted.
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(0L)
+        assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isFalse()
+        assertThat(ticker.nowMs).isEqualTo(0L)
+
+        rateLimiter.permitsPerSecond = 10L
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(1L)
+        assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isTrue()
+        assertThat(ticker.nowMs).isEqualTo(100L)
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(1L)
+    }
+
+    @Test
+    fun `rate limit decreases`() {
+        rateLimiter.permitsPerSecond = 10L
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(10L)
+        assertThat(rateLimiter.tryAcquire(10L, 100, TimeUnit.MILLISECONDS)).isTrue()
+        assertThat(ticker.nowMs).isEqualTo(0L)
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(1L)
+        assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isTrue()
+        assertThat(ticker.nowMs).isEqualTo(100L)
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(1L)
+        assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isTrue()
+        assertThat(ticker.nowMs).isEqualTo(200L)
+
+        rateLimiter.permitsPerSecond = 2L
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(0L)
+        assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isFalse()
+        assertThat(ticker.nowMs).isEqualTo(200L)
+    }
+
+    @Test
+    fun `permit count exceeds window size`() {
+        rateLimiter.permitsPerSecond = 2L
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 2_000)).isEqualTo(2L)
+        assertThat(rateLimiter.tryAcquire(3L, 2_000, TimeUnit.MILLISECONDS)).isFalse()
+        assertThat(ticker.nowMs).isEqualTo(0L)
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 2_000)).isEqualTo(2L)
+    }
+
+    @Test
+    fun `QPS is set to 0`(){
+        rateLimiter.permitsPerSecond = 0L
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 500)).isEqualTo(0L)
+        assertThat(rateLimiter.tryAcquire(1L, 500, TimeUnit.MILLISECONDS)).isFalse()
+    }
+
+    @Test
+    fun `QPS is set to a negative value`(){
+        rateLimiter.permitsPerSecond = -1L
+
+        assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 500)).isEqualTo(0L)
+        assertThat(rateLimiter.tryAcquire(1L, 500, TimeUnit.MILLISECONDS)).isFalse()
+    }
+
+    private class FakeTicker : Ticker, Sleeper {
```

Comment:
I dropped this here out of convenience.  It's the only test that needs it, but I think the more correct thing to do would have been to add a `wisp-sampling-testing` module for this, although this would mean `Ticker` and `Sleeper` could not be `internal` anymore.  Alternatively, it might fit into `wisp-time-testing` alongside `FakeClock`, but that would still require exposing the internal interfaces.

Reply:
Go with this for now, it can always be done later if we find it's useful for downstream apps to use.